src/OrchardCore/OrchardCore.Sms.Core/ServiceCollectionExtensions.cs

```diff
@@ -10,54 +11,49 @@ public static class ServiceCollectionExtensions
 {
     public static IServiceCollection AddSmsServices(this IServiceCollection services)
     {
-        services.AddSingleton<DefaultSmsProvider>();
-        services.AddTransient<IPostConfigureOptions<SmsSettings>, SmsSettingsConfigurator>();
-        services.AddSingleton(sp =>
+        services.AddTransient<IPostConfigureOptions<SmsSettings>, SmsSettingsConfiguration>();
+        services.AddSingleton(serviceProvider =>
         {
-            var settings = sp.GetRequiredService<IOptions<SmsSettings>>().Value;
+            var settings = serviceProvider.GetRequiredService<IOptions<SmsSettings>>().Value;
 
             if (!String.IsNullOrEmpty(settings.DefaultProviderName))
             {
-                var smsProviderOptions = sp.GetRequiredService<IOptions<SmsProviderOptions>>().Value;
+                var smsProviderOptions = serviceProvider.GetRequiredService<IOptions<SmsProviderOptions>>().Value;
 
                 if (smsProviderOptions.Providers.TryGetValue(settings.DefaultProviderName, out var providerGetter))
                 {
-                    return providerGetter(sp);
+                    return providerGetter(serviceProvider);
                 }
             }
 
-            return sp.GetRequiredService<DefaultSmsProvider>();
+            return serviceProvider.CreateInstance<DefaultSmsProvider>();
         });
 
         return services;
     }
 
-    public static IServiceCollection AddPhoneFormatValidator(this IServiceCollection services)
+    public static void AddPhoneFormatValidator(this IServiceCollection services)
     {
         services.TryAddScoped<IPhoneFormatValidator, DefaultPhoneFormatValidator>();
-
-        return services;
     }
 
-    public static IServiceCollection AddSmsProvider<T>(this IServiceCollection services, string name)
-        where T : class, ISmsProvider
+    public static IServiceCollection AddSmsProvider<T>(this IServiceCollection services, string name) where T : class, ISmsProvider
     {
-        services.AddSingleton<T>();
         services.Configure<SmsProviderOptions>(options =>
         {
-            options.Providers.Add(name, (sp) => sp.GetService<T>());
+            options.Providers.TryAdd(name, (sp) => sp.CreateInstance<T>());
```

Comment:
Here you can just enlist types.

Reply:
I made some changes let me know what you think. If we want to key/technical name of the provider is the class name, we can just use the name of the class and remove the `name` from being required in the registration.

Reply:
Yes looks good, I will review asap.

As you want about the key.